lib/classes/batch/class-batch-task-manager.php

```diff
@@ -0,0 +1,353 @@
+<?php
+/**
+ * Batch task manager
+ *
+ * @since 4.0.0
+ */
+
+namespace wpCloud\StatelessMedia\Batch;
+
+// Require lib classes if not yet available
+if (!class_exists('UDX_WP_Async_Request')) {
+  require_once ud_get_stateless_media()->path('lib/ns-vendor/classes/deliciousbrains/wp-background-processing/classes/wp-async-request.php', 'dir');
+}
+
+if (!class_exists('UDX_WP_Background_Process')) {
+  require_once ud_get_stateless_media()->path('lib/ns-vendor/classes/deliciousbrains/wp-background-processing/classes/wp-background-process.php', 'dir');
+}
+
+use wpCloud\StatelessMedia\Helper;
+use wpCloud\StatelessMedia\Singleton;
+
+class BatchTaskManager extends \UDX_WP_Background_Process {
+  use Singleton;
+  
+  const STATE_KEY = '_state';
+  const UPDATED_KEY = '_updated';
+  const HEALTH_CHECK_INTERVAL = 60 * 5; // 5 minute
+
+  protected $prefix = 'sm';
+  protected $action = 'batch_process';
+
+  protected function __construct() {
+    parent::__construct();
+
+    $this->_init_hooks();
+    $this->_check_force_continue();
+  }
+
+  private function _init_hooks() {
+    add_filter('wp_stateless_batch_state', [$this, 'get_state'], 10, 1);
+    add_filter('wp_stateless_batch_action_pause', [$this, 'pause_task'], 10, 2);
+    add_filter('wp_stateless_batch_action_resume', [$this, 'resume_task'], 10, 2);
+    add_filter('heartbeat_send', [$this, 'check_running_batch'], 10, 1 );
+  }
+
+  /**
+   * Check if we should force dispatch
+   * Check if the task is in progress and if the state was not updated last 5 minutes - try to continue the task
+   */
+  private function _check_force_continue() {
+    $last_updated = $this->_get_last_updated();
+
+    if ( empty($last_updated) || $this->is_paused() ) {
+      return;
+    }
+
+    $check_interval = self::HEALTH_CHECK_INTERVAL;
+
+    if ( defined('WP_STATELESS_BATCH_HEALTHCHECK_INTERVAL') ) {
+      $check_interval = max($check_interval, WP_STATELESS_BATCH_HEALTHCHECK_INTERVAL * 60);
+    }
+
+    if ( time() - $last_updated <= $check_interval ) {
+      return;
+    }
+
+    Helper::log('Batch task freezed, trying to continue...');
+
+    // Forcing continue
+    $this->unlock_process();
+    $this->handle();
+  }
+
+  /**
+   * Update current task state
+   * 
+   * @return array
+   */
+  private function _update_state($state) {
+    update_site_option( $this->identifier . self::STATE_KEY, $state );
+    update_site_option( $this->identifier . self::UPDATED_KEY, time() );
+  }
+
+  /**
+   * Get current task state
+   * 
+   * @return array
+   */
+  private function _get_state() {
+    return get_site_option( $this->identifier . self::STATE_KEY, [] );
+  }
+
+  /**
+   * Get last state update of the current task
+   * 
+   * @return array
```

Comment:
@balexey88 does it return a `string` or an `array`?
if an `array` - we have to fix line 51 and 63.
otherwise change comment here.

Reply:
Updated https://github.com/udx/wp-stateless/commit/e93b7550742f3e31acc6e30449340e30b70588b0